server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java

```diff
@@ -19,151 +19,19 @@
+public final class DocumentMapper {
 
-        public DocumentMapper build(MapperService mapperService) {
-            Objects.requireNonNull(rootObjectMapper, "Mapper builder must have the root object mapper set");
-            Mapping mapping = new Mapping(
-                    mapperService.getIndexSettings().getIndexVersionCreated(),
-                    rootObjectMapper,
-                    metadataMappers.values().toArray(new MetadataFieldMapper[metadataMappers.values().size()]),
-                    meta);
-            return new DocumentMapper(mapperService, mapping);
-        }
-    }
-
-    private final MapperService mapperService;
-
-    private final CompressedXContent mappingSource;
-
-    private final Mapping mapping;
-
-    private final DocumentFieldMappers fieldMappers;
-
-    private final Map<String, ObjectMapper> objectMappers;
-
-    public DocumentMapper(MapperService mapperService, Mapping mapping) {
-        this.mapperService = mapperService;
-        this.mapping = mapping;
-
-        // collect all the mappers for this type
-        List<ObjectMapper> newObjectMappers = new ArrayList<>();
-        List<FieldMapper> newFieldMappers = new ArrayList<>();
-        for (MetadataFieldMapper metadataMapper : this.mapping.metadataMappers) {
-            if (metadataMapper != null) {
-                newFieldMappers.add(metadataMapper);
-            }
-        }
-        MapperUtils.collect(this.mapping.root, newObjectMappers, newFieldMappers);
-
-        this.fieldMappers = new DocumentFieldMappers(newFieldMappers);
-        Map<String, ObjectMapper> builder = new HashMap<>();
-        for (ObjectMapper objectMapper : newObjectMappers) {
-            ObjectMapper previous = builder.put(objectMapper.fullPath(), objectMapper);
-            if (previous != null) {
-                throw new IllegalStateException("duplicate key " + objectMapper.fullPath() + " encountered");
-            }
-        }
-
-        this.objectMappers = Collections.unmodifiableMap(builder);
-
-        try {
-            mappingSource = new CompressedXContent(this, XContentType.JSON, ToXContent.EMPTY_PARAMS);
-        } catch (Exception e) {
-            throw new ElasticsearchGenerationException("failed to serialize source", e);
-        }
-    }
-
-    public Mapping mapping() {
-        return mapping;
-    }
-
-    public CompressedXContent mappingSource() {
-        return this.mappingSource;
-    }
-
-    public RootObjectMapper root() {
-        return mapping.root;
-    }
-
-    public <T extends MetadataFieldMapper> T metadataMapper(Class<T> type) {
-        return mapping.metadataMapper(type);
-    }
-
-    public DocumentFieldMappers mappers() {
-        return this.fieldMappers;
-    }
-
-    public Map<String, ObjectMapper> objectMappers() {
-        return this.objectMappers;
+    private DocumentMapper() {
     }
 
     public static ParsedDocument createDeleteTombstoneDoc(String index, String id) throws MapperParsingException {
```

Comment:
We could move these static methods to ParsedDocument and delete DocumentMapper entirely?